.github/workflows/build.yml

```diff
@@ -46,54 +66,148 @@ jobs:
-            .
-      - name: Save docker image artifact
+      - name: Gather repository metadata
+        id: repo
+        uses: actions/github-script@v3
+        with:
+          script: |
+            const repo = await github.repos.get(context.repo)
+            return repo.data
+      - name: Calculate output values
+        id: prep
         run: |
-          mkdir dist
-          version=${{ env.IMAGE_VERSION }}
-          docker save $IMAGE_NAME:latest | gzip > dist/image.tar.gz
+          VERSION=noop
+          SEMVER="^v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-((0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))?(\+([0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*))?$"
+          if [ "${{ github.event_name }}" = "schedule" ]; then
+            VERSION=nightly
+          elif [ "${{ github.event_name }}" = "workflow_dispatch" ]; then
+            VERSION=${{ github.event.inputs.image-tag }}
+          elif [[ $GITHUB_REF == refs/tags/* ]]; then
+            VERSION=${GITHUB_REF#refs/tags/}
+          elif [[ $GITHUB_REF == refs/heads/* ]]; then
+            VERSION=$(echo ${GITHUB_REF#refs/heads/} | sed -r 's#/+#-#g')
+            if [ "${{ github.event.repository.default_branch }}" = "$VERSION" ];
+              then
+              VERSION=edge
+            fi
+          elif [[ $GITHUB_REF == refs/pull/* ]]; then
+            VERSION=pr-${{ github.event.number }}
+          fi
+          if [[ $VERSION =~ $SEMVER ]]; then
+            VERSION_NO_V=${VERSION#v}
+            MAJOR="${BASH_REMATCH[1]}"
+            MINOR="${BASH_REMATCH[2]}"
+            PATCH="${BASH_REMATCH[3]}"
+            TAGS="${IMAGE_NAME}:${VERSION_NO_V//+/_},${IMAGE_NAME}:${MAJOR}.${MINOR}.${PATCH},${IMAGE_NAME}:${MAJOR}.${MINOR},${IMAGE_NAME}:${MAJOR},${IMAGE_NAME}:latest"
+          else
+            TAGS="${IMAGE_NAME}:${VERSION}"
+          fi
+          if [ "${{ github.event_name }}" = "push" ]; then
+            TAGS="${TAGS},${IMAGE_NAME}:sha-${GITHUB_SHA::8}"
+          fi
+          echo ::set-output name=created::$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+          echo ::set-output name=source_version::$(./bump_version.sh show)
+          echo ::set-output name=tags::${TAGS}
+          echo tags=${TAGS}
+      - name: Setup debug session remote shell
+        uses: mxschmitt/action-tmate@v3
+        if: github.event.inputs.remote-shell == 'true'
+  build:
+    name: "Build test image"
+    runs-on: ubuntu-latest
+    needs: [prepare]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.BUILDX_CACHE_DIR }}
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
```

Comment:
We've been creating keys of the form `buildx-${{ runner.os }}` for a while now.  This is a change that @mcdonnnj encouraged some time ago.